### PR TITLE
DeviceManager manages lifetimes of allocated buffers.

### PR DESCRIPTION
**Pull request by gcatron**
Summary:
This PR updates the deviceManager to manage the lifetimes of buffers allocated with allocateDeviceIOBuffer.
The buffers will be used by unowning Tensors. This way the nothing outside the DM needs to know how to free the allocated memory.
Documentation:

Test Plan:

### CODE REVIEW

File: include/glow/Backends/DeviceManager.h [hunk 1]
```diff
@@ -86,7 +90,12 @@ class DeviceManager : public DeviceTensorTransferManager {
         maxMemoryBytes_(config_.getDeviceMemory(2000000000)),
         statsExporterRegistry_(StatsExporterRegistry::Stats()) {}
 
-  virtual ~DeviceManager() = default;
+  virtual ~DeviceManager() {
+    // Free all allocated buffers.
+    for (auto pair : allocations_) {
```
Comment (jfix71) [inline include/glow/Backends/DeviceManager.h]: nit: `auto &pair`

File: include/glow/Backends/DeviceManager.h [hunk 2]
```diff
@@ -104,11 +113,21 @@ class DeviceManager : public DeviceTensorTransferManager {
...
+  virtual void *allocateDeviceIOBuffer(Placeholder *PH) {
+    void *buffer =
+        alignedAlloc(PH->getType()->getSizeInBytes(), TensorAlignment);
+    allocations_[PH].push_back(buffer);
+    return buffer;
   };
 
+  /// Free all allocated buffers associated with /p PH.
+  virtual void freeAllocatedDeviceIOBuffers(Placeholder *PH) {
+    for (auto buffer : allocations_[PH]) {
```
Comment (jfix71) [inline include/glow/Backends/DeviceManager.h]: nit: `auto *buffer`

**Comment by gcatron** [comment]
@jsubag

**Comment by yinghai** [comment]
Just to clarify, this code is not used yet, right?

**Comment by gcatron** [comment]
> Just to clarify, this code is not used yet, right?

Correct, I have a local branch that will be using this but nothing else is using it yet. Breaking this out separate for communication purposes.

**Pull request closed without merging**